src/sentry/tasks/summaries/weekly_reports.py

```diff
@@ -211,76 +224,123 @@ def prepare_organization_report(
 
     if not report_is_available:
         logger.info(
-            "prepare_organization_report.skipping_empty", extra={"organization": organization_id}
+            "prepare_organization_report.skipping_empty",
+            extra={"batch_id": batch_id, "organization": organization_id},
         )
         return
 
     # Finally, deliver the reports
+    batch = OrganizationReportBatch(ctx, batch_id, dry_run, target_user, email_override)
     with sentry_sdk.start_span(op="weekly_reports.deliver_reports"):
         logger.info("weekly_reports.deliver_reports", extra={"organization": organization_id})
-        deliver_reports(
-            ctx, dry_run=dry_run, target_user=target_user, email_override=email_override
-        )
+        batch.deliver_reports()
 
 
-def deliver_reports(
-    ctx: OrganizationReportContext,
-    dry_run: bool = False,
-    target_user: User | None = None,
-    email_override: str | None = None,
-) -> None:
-    """
-    For all users in the organization, we generate the template context for the user, and send the email.
-    """
-    # Specify a sentry user to send this email.
-    template_context: Mapping[str, Any] | None = None
-    user_id: int | None = None
-
-    if email_override:
-        target_user_id = (
-            target_user.id if target_user else None
-        )  # if None, generates report for a user with access to all projects
-        user_template_context_by_user_id_list = prepare_template_context(
-            ctx=ctx, user_ids=[target_user_id]
-        )
-        if user_template_context_by_user_id_list:
-            user_template_context_by_user_id = user_template_context_by_user_id_list[0]
-            template_context = user_template_context_by_user_id.get("context")
-            user_id = user_template_context_by_user_id.get("user_id")
-            if template_context and user_id:
-                send_email(
-                    ctx=ctx,
-                    template_ctx=template_context,
-                    user_id=user_id,
-                    dry_run=dry_run,
-                    email_override=email_override,
+@dataclass(frozen=True)
+class OrganizationReportBatch:
+    ctx: OrganizationReportContext
+    batch_id: uuid.UUID
+
+    dry_run: bool = False
+    target_user: User | None = None
+    email_override: str | None = None
+
+    def _prevent_duplicate_delivery(self) -> None:
+        from sentry.utils import redis  # TODO: Use sentry.buffer.redis instead?
+
+        cluster = redis.clusters.get("default").get_local_client_for_key("weekly_reports")
+        key = f"{self.batch_id}:{self.ctx.organization.id}"
+        count = cluster.incr(key)
+        cluster.expire(key, timedelta(days=1))
+
+        if count > 1:
+            raise Exception(f"Duplicate report delivery detected for batch {self.batch_id}")
```

Comment:
Ideally this shouldn't be hit at all. Our hypothesis is that it may be hit regardless. Raising an exception would prevent users from getting spammed with extra messages until we deliver a fix.

However, it incurs some risk that it could incorrectly prevent the *first* delivery. If we want to be more cautious, we could downgrade it to an error log and see if it gets hit at all.